src/Geta.Optimizely.Categories/CategoryCriterion.cs

```diff
@@ -51,35 +44,89 @@ public override void Unsubscribe(ICriterionEvents criterionEvents)
+            }
+
+            if (!int.TryParse(Model.CategoryId, out var categoryId))
             {
                 return;
             }
-            var pageCatIds = categorizable.Categories?.Select(x => x.ID);
-            if (_stateStorage.IsAvailable && pageCatIds != null && pageCatIds.Contains(int.Parse(Model.CategoryId)))
+
+            var viewedCategories = Load(e.HttpContext);
+
+            // When already true, no need to update anymore
+            if (GetVisitedTimes(viewedCategories) >= Model.ViewedTimes)
+            {
+                return;
+            }
+
+            if (!_contentLoader.TryGet<IContent>(e.GetPageLink(), out var page)
+                || page is not ICategorizableContent categorizable)
+            {
+                return;
+            }
+
+            // Test if it matches the configured cat id
+            var pageCatIds = categorizable.Categories?.Select(x => x.ID) ?? Enumerable.Empty<int>();
+            if (!pageCatIds.Contains(categoryId))
             {
-                var times = GetVisitedTimes() + 1;
-                _stateStorage.Save(_STORAGEKEY, times);
+                return;
             }
+
+            var count = viewedCategories.ContainsKey(categoryId) ? viewedCategories[categoryId] : 0;
+            viewedCategories[categoryId] = ++count;
+            Save(e.HttpContext, viewedCategories);
         }
 
-        private int GetVisitedTimes()
+        private IDictionary<int, int> Load(HttpContext httpContext)
         {
-            var timesObj = _stateStorage.Load(_STORAGEKEY);
-            if (timesObj == null || string.IsNullOrEmpty(timesObj.ToString()))
+            try
+            {
+                var data = httpContext?.Items?[STORAGEKEY] as string ?? _stateStorage.Load(STORAGEKEY) as string;
+                if (data != null)
+                {
+                    return JsonSerializer.Deserialize<Dictionary<int, int>>(data);
+                }
+            }
+            catch (Exception e)
             {
-                return 0;
+                _logger.LogError(e, "Could not load Geta Category criterion from storage");
             }
+            return new Dictionary<int, int>();
+        }
 
-            if (int.TryParse(timesObj.ToString(), out int times))
+        private void Save(HttpContext httpContext, IDictionary<int, int> viewedCategories)
+        {
+            if (httpContext?.Items != null)
             {
-                return times;
+                httpContext.Items[STORAGEKEY] = viewedCategories;
             }
+            _stateStorage.Save(STORAGEKEY, JsonSerializer.Serialize(viewedCategories));
+        }
 
-            return 0;
+        private int GetVisitedTimes(IDictionary<int, int> viewedCategories)
+        {
+            return viewedCategories != null
+                   && int.TryParse(Model.CategoryId, out var cat)
+                   && viewedCategories.TryGetValue(cat, out var count) ?
+                count :
+                0;
```

Comment:
I would format it a little bit differently so it is easier to read:
```suggestion
                   && viewedCategories.TryGetValue(cat, out var count)
                   ? count
                   : 0;
```